teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,190 @@
+2. The amount of attempts
+3. Actions to take after a failed attempt
+4. Timeout of each attempt
+
+| | [Retry Step in GA](https://github.com/marketplace/actions/retry-step) | [Retry Action in GA](https://github.com/marketplace/actions/retry-action) | [Argo Workflow Template](https://argoproj.github.io/argo-workflows/retries/#retry-policies) | [GitLab Job](https://docs.gitlab.com/ee/ci/yaml/#retry) | [Ansible Task](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#retrying-a-task-until-a-condition-is-met)| [Concourse Step](https://concourse-ci.org/attempts-step.html#attempts-step) |
+|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|
+
+Several observations regarding to the feature table above:
+- We can configure **timeout duration per attempt** in all CI systems that support the `retry` functionality.
+- GitHub Action doesn't support retry natively, but because the flexibility of **customized actions**, some users write their own `retry` action to make it work, and those customized actions even support what to do before retrying a failed attempt.
+- Concourse mentioned the reason it [retries per attempt is somewhat arbitrary](https://concourse-ci.org/attempts-step.html#attempts-step).
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Options Under Consideration
+
+No matter how we implement the retry functionality, we propose to set `Timeout` for each retry attempt. This is an industrial standard per [related work](#related-work).
+
+### Option 1: Implement `retries` in TaskRun
+- Stop relying on `len(retriesStatus)` to determine whether a TaskRun or CustomRun finishes, use `ConditionSucceeded` & `ConditionFalse` instead.
+- `Retries` and `Timeout` are passed from `PipelineTask` to `TaskRunSpec` and `CustomRunSpec`.
+
+Two sub-options about the way to implement `retriesStatus`:
+- 1.a: Update `retriesStatus` for each retry attempt for `TaskRun`.
+  - No API deprecation.
+- 1.b: Deprecate `retriesStatus` for both `TaskRun` and `CustomRun`, create a new `TaskRun` for each retry attempt.
```

Comment:
I don't think it would make sense for a TaskRun to create a new TaskRun to retry-- if we implement retries in TaskRun it would probably make sense for TaskRun to still have retriesStatus. I think option 1.b here is actually just deprecating retriesStatus in CustomRun.

Reply:
Thanks for the comment Lee, added it as a third option for discussion.